### PR TITLE
[SYCL] Implement sycl_ext_oneapi_kernel_compiler_spirv

### PR DESCRIPTION
**Pull request by 0x12CC**
Implements the extension described in #11954. This PR includes the following changes:
- Adds a `create_kernel_bundle_from_source` overload for `std::vector<std::byte>` kernel sources.
- Adds new `source_language::spirv`.
- Defines `SYCL_EXT_ONEAPI_KERNEL_COMPILER_SPIRV`.
- Adds support for SPIR-V kernels created from `std::vector<std::byte>` sources.
- Moves `sycl_ext_oneapi_kernel_compiler_spirv.asciidoc` from `proposed` to `experimental`.

### CODE REVIEW

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 1]
```diff
@@ -334,6 +335,14 @@ class kernel_bundle_impl {
       : MContext(Context), MDevices(Context.get_devices()),
         MState(bundle_state::ext_oneapi_source), Language(Lang), Source(Src) {}
 
+  // oneapi_ext_kernel_compiler
+  // construct from source bytes
+  kernel_bundle_impl(const context &Context, syclex::source_language Lang,
+                     const std::vector<std::byte> &Bytes)
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/source/detail/kernel_bundle_impl.hpp:341]: It looks like our convention is to put `#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)` around any SYCL API that uses `std::byte`, making the API unavailable on MSVC when `std::byte` support is disabled.  I guess we should do the same for this API.

File: sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  // Build the SPIR-V module for our device.
+  sycl::kernel_bundle<sycl::bundle_state::executable> kb_exe =
+      syclex::build(kb_src);
+
+  // Get a "kernel" object representing the kernel from the SPIR-V module.
+  sycl::kernel my_kernel = kb_exe.ext_oneapi_get_kernel("my_kernel");
+
+  // Test the kernel
+  auto my_num_args = my_kernel.get_info<sycl::info::kernel::num_args>();
+  assert(my_num_args == 2 && "my_kernel should take 2 args");
```
Comment (gmlueck) [inline sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:72]: This test for "num_args" is kind of confusing.  A reader would probably think that the two arguments are the `2` and the `100` that you pass below, but those are totally unrelated to the kernel arguments.

I think it might be clearer to move this test into `testSyclKernel`, where the reader can see that the kernel really does have 2 arguments.

File: sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp
Comment (gmlueck) [inline sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:1]: I think we should have tests for all the kernel argument types defined in this section of the extension spec:

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler_spirv.asciidoc#passing-kernel-arguments

I think the best way to do this is to write the kernel in SPIR-V instead of OpenCL C.  (You could start with an OpenCL C kernel and compile it to SPIR-V in order to get the general framework.  Then you can edit the SPIR-V file to make sure the details are right.)

The test should include the following kernel argument types:

* **OpTypeInt**, width 8, 16, 32, and 64.
* **OpTypeFloat**, width 16, 32 and 64.
* **OpTypePointer** with storage class **CrossWorkgroup**, pass argument as USM pointer.
* **OpTypePointer** with storage class **CrossWorkgroup**, pass argument as `accessor`.
* **OpTypePointer** with storage class **Workgroup**, pass argument as `local_accessor`.
* **OpTypeStruct** with members **OpTypeInt**, **OpTypeFloat**, and  **OpTypePointer** (storage class **CrossWorkgroup**)

Note that **OpTypeFloat** with widths 16 and 64 are optional features.  Therefore, the test will need to check the device aspects `aspect::fp16` and `aspect::fp64` before using these argument types.

File: sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  spv_stream.read(reinterpret_cast<char *>(spv.data()), sz);
+
+  // Create a kernel bundle from the binary SPIR-V.
+  sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src =
+      syclex::create_kernel_bundle_from_source(
+          q.get_context(), syclex::source_language::spirv, spv);
+
+  // Build the SPIR-V module for our device.
+  sycl::kernel_bundle<sycl::bundle_state::executable> kb_exe =
+      syclex::build(kb_src);
```
Comment (cperkinsintel) [inline sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:45]: `sycl::build` also takes a property list for logging and build options. Are those supported by SPIR-V as well? If so we should test them.

File: sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp
Comment (gmlueck) [reply sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:1]: > For the `OpTypeStruct` test case, I can't use the same approach. The struct parameter in the SPIR-V binary was represented as an `OpTypePointer` to `OpTypeStruct`. Also, OpenCL C disallows structs containing pointers.

Yes, good catch.  I did some experimentation, and I agree that **OpTypeStruct** is not a valid kernel argument type.  As you say, it should be **OpTypePointer** (with _Storage Class_ of **Function**), where the pointer points to an **OpTypeStruct**.  I'll update the extension specification to fix this.

> Could you please clarify this suggestion? If I manually modify the generated binary, I would need to add it to this PR.

Yes, this is what I had in mind.  It seems better for the test to include the SPIR-V file, rather than assuming the OpenCL C compiler generates a certain SPIR-V.  I thought that this PR should include assembly-form SPIR-V modules as part of the test.  The test can then use `spirv-as` to translate the assembly code into binary.

FWIW, it seems that Level Zero only supports SPIR-V 1.2, so you need to run `spirv-as` with the option `--target-env spv1.2`.

Comment (gmlueck) [reply sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:1]: FYI: #12338 updates the spec regarding **OpTypeStruct**.

File: sycl/include/sycl/kernel_bundle.hpp [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 #pragma once
 
+#include <cstddef>                         // for std::byte
```
Comment (AlexeySachkov) [inline sycl/include/sycl/kernel_bundle.hpp:11]: Should be moved _after_ all SYCL includes, so that they don't accidentally rely on it being implicitly included

File: sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+void testKernelsFromSpvFile(std::string kernels_file,
+                            std::string fp16_kernel_file,
+                            std::string fp64_kernel_file) {
+  const auto getKernel =
+      [](sycl::kernel_bundle<sycl::bundle_state::executable> &bundle,
+         const std::string &name) {
+        return bundle.ext_oneapi_get_kernel(name);
+      };
+
+  sycl::queue q;
```
Comment (cperkinsintel) [inline sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:119]: if you are having test failures on PVC, it is because there is a bug where this gets set up with multiple devices, not just one.  L0 will then refuse to instantiate the zeModule/zeProgram  ("one device max" or some such)
The workaround is to do this:
```
  sycl::device d;
  sycl::context ctx{d};
  sycl::queue q{ctx, d};
```

until that gets fixed.

File: sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
+//==- kernel_compiler_spirv.cpp --------------------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: ocloc
```
Comment (cperkinsintel) [inline sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:9]: Just a quick note, I made a normal SYCL app with a simple kernel and used `-fsycl-dump-device-code=./` when compiling it, then I fed that same .spv file to your kernel_bundle, was able to call `has_kernel(name)` and `get_kernel(name)` and pass it successfully to `parallel_for` . So that could be another way of testing, and without the ocloc dependency.

**Review by bso-intel — APPROVED** [review]

File: sycl/include/sycl/kernel_bundle.hpp [hunk 7]
```diff
@@ -886,11 +887,18 @@ __SYCL_EXPORT bool is_source_kernel_bundle_supported(backend BE,
...
+
+#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
 __SYCL_EXPORT kernel_bundle<bundle_state::ext_oneapi_source>
-create_kernel_bundle_from_source(
-    const context &SyclContext,
-    sycl::ext::oneapi::experimental::source_language Language,
-    const std::string &Source);
+create_kernel_bundle_from_source(const context &SyclContext,
+                                 source_language Language,
+                                 const std::vector<std::byte> &Bytes);
```
Comment (cperkinsintel) [inline sycl/include/sycl/kernel_bundle.hpp:900]: @gmlueck I definitely think creating kernel bundles from spirv binary ( .spv ) is the right thing.  

But there is a tiny bit of a disconnect there, where we are creating "from source" and yes that source is .spv binary not .spt the textual human(*) readable representation of SPIR-V.  Anyway, I'm not trying to make more work for anyone. Nor do I think there is any real demand for using .spt sources.  But it does seem a bit funny.

(*) I guess .spt is "almost human" readable.

File: sycl/source/kernel_bundle.cpp [hunk 8]
```diff
@@ -372,7 +374,9 @@ bool is_source_kernel_bundle_supported(backend BE, source_language Language) {
   // and it's support is limited to the opencl and level_zero backends.
```
Comment (cperkinsintel) [inline sycl/source/kernel_bundle.cpp:374]: Comment for "at the moment OpenCL is the only language supported" should be updated.

Right now these two "acceptable" backends is because they support spirv kernel creation and runtime compilation. CUDA and others are AoT only. Maybe we should put that in the comment too?

File: sycl/source/kernel_bundle.cpp [hunk 9]
```diff
@@ -372,7 +374,9 @@ bool is_source_kernel_bundle_supported(backend BE, source_language Language) {
...
   bool BE_Acceptable = (BE == sycl::backend::ext_oneapi_level_zero) ||
                        (BE == sycl::backend::opencl);
-  if ((Language == source_language::opencl) && BE_Acceptable) {
+  if ((Language == source_language::opencl ||
+       Language == source_language::spirv) &&
+      BE_Acceptable) {
     return detail::OpenCLC_Compilation_Available();
```
Comment (cperkinsintel) [inline sycl/source/kernel_bundle.cpp:380]: Why does SPIR-V language support need "OpenCLC_Compiation_Available()"  ? I think this function should probably be re-written with the addition of SPIRV support

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 10]
```diff
@@ -350,17 +359,30 @@ class kernel_bundle_impl {
                     std::string *LogPtr) {
     assert(MState == bundle_state::ext_oneapi_source &&
            "bundle_state::ext_oneapi_source required");
-    assert(Language == syclex::source_language::opencl &&
-           "TODO: add other Languages. Must be OpenCL");
-    if (Language != syclex::source_language::opencl)
+    assert((Language == syclex::source_language::opencl ||
```
Comment (cperkinsintel) [inline sycl/source/detail/kernel_bundle_impl.hpp:362]: I think I was mistaken when I originally put this assert in here. We throw an exception if the language isn't supported. And `language` is an enumerated type, so it's not really possible for it to be any other value.  I think this whole assertion of language value is redundant and should be removed.

File: sycl/include/sycl/kernel_bundle.hpp (hunk 7 above)
Comment (gmlueck) [reply sycl/include/sycl/kernel_bundle.hpp:900]: I agree that calling SPIR-V binary "source" is a bit of a misnomer, but all the APIs work out pretty well.  We could easily add SPIR-V source support in the future by adding a new enumerator like `source_language::spirv_source`.  However, I think there is no compelling reason to do that unless there is some request.

**Review by cperkinsintel — APPROVED** [review]

File: sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp
Comment (gmlueck) [reply sycl/test-e2e/KernelCompiler/kernel_compiler_spirv.cpp:1]: This is OK with me, but I'm not sure if we have a policy against adding binary files to the repo.  I'll let the other reviewers comment on that.

If someone thinks we could somehow add the [SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools) repo as a dependency, I think that would be a cleaner solution.

**Review by gmlueck — APPROVED** [review]
